File: app/src/main/java/com/android/virgilsecurity/virgilback4app/auth/LogInPresenter.java
package com.android.virgilsecurity.virgilback4app.auth;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Base64;

import com.android.virgilsecurity.virgilback4app.AppVirgil;
import com.android.virgilsecurity.virgilback4app.util.InfoHolder;
import com.android.virgilsecurity.virgilback4app.util.RxParse;
import com.virgilsecurity.sdk.client.exceptions.VirgilKeyIsAlreadyExistsException;
import com.virgilsecurity.sdk.client.exceptions.VirgilKeyIsNotFoundException;
import com.virgilsecurity.sdk.crypto.exceptions.CryptoException;
import com.virgilsecurity.sdk.crypto.exceptions.KeyEntryNotFoundException;
import com.virgilsecurity.sdk.highlevel.VirgilCard;
import com.virgilsecurity.sdk.highlevel.VirgilKey;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import nucleus5.presenter.RxPresenter;

/**
 * Created by Danylo Oliinyk on 11/17/17 at Virgil Security.
 * -__o
 */

public class LogInPresenter extends RxPresenter<LogInFragment> {

    private static final int SIGN_UP = 0;
    private static final int LOG_IN = 1;

    private InfoHolder infoHolder;

    private VirgilCard virgilCard;
    private VirgilKey virgilKey;

    private String identity;
    private String password;

    @Override
    protected void onCreate(@Nullable Bundle savedState) {
        super.onCreate(savedState);

        infoHolder = AppVirgil.getInfoHolder();

        restartableFirst(SIGN_UP, () ->
                                 RxParse.signUp(identity,
                                                password,
                                                virgilCard)
                                        .subscribeOn(Schedulers.io())
                                        .observeOn(AndroidSchedulers.mainThread()),
                         LogInFragment::onSignUpSuccess,
                         LogInFragment::onSignUpError
        );

        restartableFirst(LOG_IN, () ->
                                 RxParse.logIn(identity,
                                               password)
                                        .subscribeOn(Schedulers.io())
                                        .observeOn(AndroidSchedulers.mainThread()),
                         LogInFragment::onLoginSuccess,
                         LogInFragment::onLoginError
        );
    }

    void requestSignUp(String identity) {
        this.identity = identity;

        generateKeyPair(identity);
        password = generatePassword(virgilKey.getPrivateKey().getValue());

        start(SIGN_UP);
    }

    void requestLogIn(String identity) {
        this.identity = identity;

        try {
            virgilKey = infoHolder.getVirgilApi().getKeys().load(identity);
        } catch (KeyEntryNotFoundException e) {
            getView().onLoginError(e);
            return;
        } catch (VirgilKeyIsNotFoundException e) {
            getView().onLoginError(e);
            return;
        } catch (CryptoException e) {
            getView().onLoginError(e);
            return;
        }
        password = generatePassword(virgilKey.getPrivateKey().getValue());

        start(LOG_IN);
    }

    void disposeAll() {
        stop(SIGN_UP);
        stop(LOG_IN);
    }

    boolean isDisposed() {
        return isDisposed(SIGN_UP)
                || isDisposed(LOG_IN);
    }

    /**
     * Generates SHA-256 password from base[]
     *
     * @param privateKey byte[] representation of private key
     * @return
     */
    private static String generatePassword(byte[] privateKey) {
        MessageDigest sha;
        byte[] hash = new byte[0];

        try {
            sha = MessageDigest.getInstance("SHA-256");
            hash = sha.digest(privateKey);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }

        return Base64.encodeToString(hash, Base64.DEFAULT);
    }

    private void generateKeyPair(String identity) {
        virgilKey = infoHolder.getVirgilApi().getKeys().generate();
        try {
            virgilKey.save(identity);
        } catch (VirgilKeyIsAlreadyExistsException e) {
            e.printStackTrace();
        }
        virgilCard = infoHolder.getVirgilApi().getCards().create(identity, virgilKey);
    }
}

File: app/src/main/java/com/android/virgilsecurity/virgilback4app/auth/LogInPresenter.kt
package com.android.virgilsecurity.virgilback4app.auth

import android.content.Context
import android.util.Base64
import com.android.virgilsecurity.virgilback4app.AppVirgil
import com.android.virgilsecurity.virgilback4app.util.AuthRx
import com.android.virgilsecurity.virgilback4app.util.Preferences
import com.android.virgilsecurity.virgilback4app.util.RxEthree
import com.android.virgilsecurity.virgilback4app.util.RxParse
import com.parse.ParseUser
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * Created by Danylo Oliinyk on 11/17/17 at Virgil Security.
 * -__o
 */

class LogInPresenter(context: Context) {

    private val compositeDisposable = CompositeDisposable()
    private val rxEthree = RxEthree(context)
    private val preferences = Preferences.instance(context)

    fun requestSignUp(identity: String, onSuccess: () -> Unit, onError: (Throwable) -> Unit) {
        val password = generatePassword(identity.toByteArray())

        val disposable = RxParse.signUp(identity, password)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .toSingle { ParseUser.getCurrentUser() }
                .flatMap { AuthRx.virgilJwt(it.sessionToken) }
                .map { preferences.setVirgilToken(it) }
                .flatMap { rxEthree.initEthree(identity) }
                .map { AppVirgil.eThree = it }
                .flatMap { rxEthree.registerEthree().toSingle { Unit } }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onSuccess = {
                        onSuccess()
                    },
                    onError = {
                        onError(it)
                    }
                )

        compositeDisposable += disposable
    }

    /**
     * Sign In user in Back4App, exchange Back4App User's sessionToken for Virgil Jwt,
     * then initialize and register [EThree].
     */
    fun requestSignIn(identity: String,
                      onSuccess: () -> Unit,
                      onError: (Throwable) -> Unit) {

        val password = generatePassword(identity.toByteArray())

        val disposable = RxParse.logIn(identity, password)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .flatMap { AuthRx.virgilJwt(it.sessionToken) }
                .map { preferences.setVirgilToken(it) }
                .flatMap { rxEthree.initEthree(identity, true) }
                .map { AppVirgil.eThree = it }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onSuccess = {
                        onSuccess()
                    },
                    onError = {
                        onError(it)
                    }
                )

        compositeDisposable += disposable
    }

    internal fun disposeAll() {
        compositeDisposable.clear()
    }

    /**
     * !!! NOT SECURE !!!
     *
     * It's just for example purpose. Ask your user for the password.
     *
     * This function generates SHA-256 password as Base64 String from provided data.
     */
    private fun generatePassword(data: ByteArray): String {
        val sha: MessageDigest
        val hash: ByteArray

        try {
            sha = MessageDigest.getInstance("SHA-256")
            hash = sha.digest(data)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            throw IllegalStateException("No Such Algorithm")
        }

        return Base64.encodeToString(hash, Base64.DEFAULT)
    }
}
